src/coreclr/jit/importer.cpp

```diff
@@ -11309,13 +11309,35 @@ var_types Compiler::impGetByRefResultType(genTreeOps oper, bool fUnsigned, GenTr
 
         if (genActualType(op1->TypeGet()) != TYP_I_IMPL)
         {
-            // insert an explicit upcast
+// insert an explicit upcast
+#ifdef TARGET_LOONGARCH64
```

Comment:
// For LoongArch64's instructions operation of the 64bits and 32bits using the whole
                // 64bits-width register which is unlike the AMD64 and ARM64.
                // And the INT type instruction will be signed-extend by default.
                // e.g. 'ld_w $r4, $5, 4' and `addi_w $r4,$r5,-1` the result of INT
                // will be signed-extend by default.

For the const-tree, there is no need adding a new tree node for casting.

Reply:
> If I am reading this code correctly, it is creating `TYP_INT` operands mixed in with `TYP_LONG` operators, e. g. `ADD long(int, int)`.
> 
> This is not valid IR as-is. I think altering the semantics of IR (**especially** HIR) like this only for LA is highly undesirable, and any such alterations should be confined to the LA backend only (preferably codegen, because it is likewise desirable for basic IR shapes to have the same semantics in both LIR and HIR).

Yeh, it's very difficult to alterate the TYP_INT just for liking LA and might be undesirable at least right now.
Of course here we didn't want or require to do it.

Let's get to the point, here is just for the const-value, maybe for the const value there is no need to do some type-casting within integer type, in fact we can do it within the C++ instead of by JIT-compiling-instructions.

Reply:
>  Of course here we didn't want or require to do it.

That's very reassuring to hear! I believe if you remove this type mixing, no frontend changes that @BruceForstall also commented on earlier (and you have added the comment above for) will be needed.

> Let's get to the point, here is just for the const-value, maybe for the const value there is no need to do some type-casting within integer type, in fact we can do it within the C++ instead of by JIT-compiling-instructions.

Right, but that is an IR size optimization that is best done separately from this change. `CAST(CNS_INT)` will be folded in later phases (morph), so you should not see it in codegen, unless optimizations are disabled (in which case it is by design that the casts remain).

Note that as-is this code omits the cast to `TYP_LONG`:
```c
 #ifdef TARGET_LOONGARCH64
if (op1->TypeGet() == TYP_INT && op1->gtOper == GT_CNS_INT)
{
    op1->AsIntCon()->gtIconVal =
        fUnsigned ? (uint32_t)op1->AsIntCon()->gtIconVal : op1->AsIntCon()->gtIconVal;
    op1->gtType = TYP_LONG;
}
else if (op1->gtOper == GT_CNS_INT)
    // This is unreachable.
    *pOp1 = gtNewCastNode(TYP_I_IMPL, op1, fUnsigned, fUnsigned ? TYP_U_IMPL : TYP_I_IMPL);
#else
```

Reply:
> Less tree nodes is better and optimize should be earlier.

As I noted, I fully agree that we should not be creating these casts from constants. It has actually long been on my `TODO-Casts` backlog. I just think it should not be part of this change, and should not be arch-specific (TP gains would apply equally everywhere).

And regardless of that, we should not be creating invalid IR.

Reply:
> And regardless of that, we should not be creating invalid IR.

Agree with you.
Thanks

And we plan to submit the optimization seperatly in future after we finish pushing the LA into main.